src/Database/LSMTree/Model/Normal/Admin.hs

```diff
@@ -0,0 +1,383 @@
+  , snapshots         :: Map SUT.SnapshotName Snapshot
+  }
+  deriving Show
+
+initModel :: Model
+initModel = Model {
+      tableHandles = Map.empty
+    , nextTableHandleID = 0
+    , snapshots = Map.empty
+    }
+
+data SomeTable where
+  SomeTable :: forall k v blob. C k v blob
+            => Model.Table k v blob -> SomeTable
+
+instance Show SomeTable where
+  show (SomeTable table) = show table
+
+--
+-- Constraints
+--
+
+-- | Common constraints for keys, values and blobs
+type C k v blob = (
+    Show k, Show v, Show blob
+  , Eq k, Eq v, Eq blob
+  , Typeable k, Typeable v, Typeable blob
+  )
+
+--
+-- ModelT and ModelM
+--
+
+newtype ModelT m a = ModelT { _runModelT :: ExceptT Err (StateT Model m) a }
+  deriving stock Functor
+  deriving newtype ( Applicative
+                   , Monad
+                   , MonadState Model
+                   , MonadError Err
+                   )
+
+runModelT :: ModelT m a -> Model -> m (Either Err a, Model)
+runModelT = runStateT . runExceptT . _runModelT
+
+type ModelM = ModelT Identity
+
+runModelM :: ModelM a -> Model -> (Either Err a, Model)
+runModelM m = runIdentity . runModelT m
+
+--
+-- Errors
+--
+
+data Err =
+    ErrTableHandleClosed TableHandleID
+  | ErrSnapshotExists TableHandleID SUT.SnapshotName
+  | ErrSnapshotDoesNotExist SUT.SnapshotName
+  | ErrSnapshotWrongType SUT.SnapshotName
+  deriving (Show, Eq)
+
+{-------------------------------------------------------------------------------
+  Tables
+-------------------------------------------------------------------------------}
+
+type TableHandleID = Int
+
+--
+-- API
+--
+
+data TableHandle k v blob = TableHandle {
+    tableHandleID :: TableHandleID
+  , config        :: SUT.TableConfig
+  }
+  deriving Show
+
+new ::
+     forall k v blob m. (MonadState Model m, C k v blob)
+  => SUT.TableConfig
+  -> m (TableHandle k v blob)
+new config = state $ \Model{..} ->
+    let tableHandle = TableHandle {
+            tableHandleID = nextTableHandleID
+          , ..
+          }
+        someTable = SomeTable $ Model.empty @k @v @blob
+        tableHandles' = Map.insert nextTableHandleID someTable tableHandles
+        nextTableHandleID' = nextTableHandleID + 1
+        model' = Model {
+            tableHandles = tableHandles'
+          , nextTableHandleID = nextTableHandleID'
+          , ..
+          }
+    in  (tableHandle, model')
+
+-- |
+--
+-- This is idempotent.
+close :: MonadState Model m => TableHandle k v blob -> m ()
+close TableHandle{..} = state $ \Model{..} ->
+    let tableHandles' = Map.delete tableHandleID tableHandles
+        model' = Model {
+            tableHandles = tableHandles'
+          , nextTableHandleID
+          , ..
+          }
+    in ((), model')
+
+--
+-- Utility
+--
+
+guardTableHandleIsClosed ::
+     forall k v blob m. (MonadState Model m, MonadError Err m)
+  => TableHandle k v blob
+  -> m (Model.Table k v blob)
+guardTableHandleIsClosed TableHandle{..} =
```

Comment:
Naming: `guard` asserts positive conditions, so this should be named `guardTableHandleIsOpen` or `NotClosed`.